### DISCUSSION TITLE
[2일차 -윤경] 브라우저의 동작원리

### DISCUSSION THREAD

**Discussion by skyhanull** [zzangCS/zzangInterview#11] [프론트 면접 질문]
브라우저가 동작하는 원리에대해 설명하기

**Comment by smarfy99** [comment]
https://velog.io/@sylagape1231/%EC%A3%BC%EC%86%8C%EC%B0%BD%EC%97%90-naver.com%EC%9D%84-%EC%B9%98%EB%A9%B4-%EC%9D%BC%EC%96%B4%EB%82%98%EB%8A%94-%EC%9D%BC%EC%9D%84-%EC%89%BD%EA%B2%8C-%EC%9D%B4%ED%95%B4%ED%95%B4%EB%B3%B4%EC%9E%90
읽고 추후 제대로 정리

**Comment by skyhanull** [comment]
1.HTML를 파싱 후, DOM트리를 구축합니다
2.CSS를 파싱 후, CSSOM트리를 구축합니다
3.Javascript를 실행합니다 
4.DOM과 CSSOM을 조합하여 렌더트리를 구축합니다.
5.뷰포트 기반으로 렌더트리의 각 노드가 가지는 정확한 위치와 크기를 계산합니다(레이아웃 단계)
6.계산한 위치/크기를 기반으로 화면에 그립니다.(페인트 단계)

**Comment by ez1n** [comment]
- **브라우저**
> 동기적으로 HTML, CSS, JS 언어를 해석해서 내용을 화면에 보여주는 응용 소프트웨어

- **브라우저 구조**
1. 사용자 인터페이스 : 사용자가 접근할 수 있는 영역
2. 브라우저 엔진 : UI와 렌더링 엔진 사이의 동작 제어
3. 렌더링 엔진 : 웹 서버로부터 응답받은 리소스를 웹 브라우저 상에 나타내는 기능
 ⇒ 서버로부터 HTML문서를 응답받으면 렌더링 엔진의 HTML 파서와 CSS 파서에 의해 파싱되어 DOM, CSSOM 트리로 변환되고 렌더 트리로 결합된다.
4. 통신 : HTTP 요청과 같은 네트워크호출에 사용
5. UI 백엔드 : 기본적인 장치를 그림
6. 자바스크립트 해석기 : JS 코드를 해석하고 실행
7. 자료 저장소 : 브라우저 메모리를 활용하여 저장하는 영역

- **렌더링 엔진 동작 과정**
1. HTML문서를 파싱하여 DOM 트리 구축
2. CSS 파일과 함께 포함된 스타일 요소 파싱 후 DOM 트리와 함께 렌더 트리 구축
3. 렌더 트리 배치
4. UI 백엔드에서 렌더 트리 그리기

```
💡 **자바스크립트**

자바스크립트는 자바스크립트 엔진이 처리하기 때문에 HTML 파서는 script 태그를 만나면 JS 코드 실행을 위해 DOM 생성 프로세스를 중지하고 자바스크립트 엔진으로 권한을 넘긴다.

이후 자바스크립트 엔진은 script 태그 내의 JS 코드나 src 속성에 정의된 JS 파일을 로드해 파싱 후 실행한다.

실행이 완료되면 다시 HTML 파서로 제어 권한을 넘겨 중지했던 시점으로 돌아가 DOM 생성을 재개한다!
```


- **브라우저의 동작 원리**
1. Loading : HTTP 모듈 또는 파일 시스템으로 전달받은 리소스를 읽는 과정
2. DOM tree 빌드 : HTML 마크업 처리 (무엇을 그릴지)
3. CSSOM Tree 빌드 : CSS 마크업 처리 (어떻게 그릴지)
4. 렌더링 트리 형성 : 생성한 DOM 및 CSSOM 트리 결합 (화면에 그려질 것이 무엇인지)
5. Layout 과정 : 렌더링 트리에서 각 노드의 형태를 계산하여 Box-Model 생성
6. Paint 과정 : 개별 노드를 화면에 그림

**Reply by ez1n** [reply]
https://d2.naver.com/helloworld/59361

**Reply by ez1n** [reply]
CSSOM
브라우저는 html을 DOM으로 만들면서 개발자가 정의해놓은 CSS와 기본적으로 설정되어있는 CSS를 cascading rule에 따라서 병합하여 CSSOM을 만들게 된다.
